### PR TITLE
provide API to allow concurrent modification of arrays, and CAS operations on fields, with JVM version dependent implementations

### PR DESCRIPTION
**Pull request by richardstartin**
This provides an API for modifying array contents concurrently, using `VarHandle` on JDK9+ and `sun.misc.Unsafe` before that was available (and before the illegal reflective access was logged).

### CODE REVIEW

File: internal-api/src/main/java/datadog/trace/api/Platform.java [hunk 1]
```diff
@@ -30,4 +29,109 @@ static int parseJavaVersion(String javaVersion) {
...
+              loadAndInstantiate(
+                  "datadog.trace.core.varhandles.VarHandleConcurrentArrayOperations");
+        } else {
+          operations =
+              loadAndInstantiate("datadog.trace.core.unsafe.UnsafeConcurrentArrayOperations");
+        }
+      } catch (Throwable t) {
+        // is it safe to log here? Weird visibility bugs will likely
+        // follow from this, but the tracer will more or less function
+        // as expected
```
Comment (bantonsson) [inline internal-api/src/main/java/datadog/trace/api/Platform.java:120]: I can't see why it wouldn't be ok to log here, and I think that there needs to be something logged. As you say in the comment, the `NonConcurrentArrayOperationsFallback` can't really guarantee the visibility of the API.

**Review by bantonsson — APPROVED** [review]
Only minor comment about logging, and I hope that the API is version independent 😉

File: internal-api/src/main/java/datadog/trace/api/Platform.java (hunk 1 above)
Comment (richardstartin) [reply internal-api/src/main/java/datadog/trace/api/Platform.java:120]: Yes I agree

**Comment by richardstartin** [comment]
The API is version independent, but the implementation loaded depends on the JVM. Ideally we'd be able to use Multi-Release jars for this, but can't.

File: internal-api/src/main/java/datadog/trace/unsafe/ConcurrentArrayOperations.java [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+public interface ConcurrentArrayOperations {
+
+  Object getObjectVolatile(Object[] array, int index);
+
+  void putObjectVolatile(Object[] array, int index, Object x);
+
+  int getIntVolatile(int[] array, int index);
+
+  long getLongVolatile(long[] array, int index);
```
Comment (mcculls) [inline internal-api/src/main/java/datadog/trace/unsafe/ConcurrentArrayOperations.java:11]: can you move this down to just above `putLongVolatile`, to keep the method ordering consistent

File: utils/varhandles/src/main/java/datadog/trace/core/varhandles/VarHandleConcurrentArrayOperations.java [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+  }
+
+  @Override
+  public void putObjectVolatile(Object[] array, int index, Object x) {
+    OBJECT_ARRAY_VAR_HANDLE.setVolatile(array, index, x);
+  }
+
+  @Override
+  public int getIntVolatile(int[] array, int index) {
+    return (int) INT_ARRAY_VAR_HANDLE.getAcquire(array, index);
```
Comment (mcculls) [inline utils/varhandles/src/main/java/datadog/trace/core/varhandles/VarHandleConcurrentArrayOperations.java:30]: is there a reason this isn't `getVolatile`?

Comment (richardstartin) [reply utils/varhandles/src/main/java/datadog/trace/core/varhandles/VarHandleConcurrentArrayOperations.java:30]: No, more haste less speed...

**Comment by richardstartin** [comment]
I will put this on hold for the time being, since building a module with JDK11 features is a non-trivial exercise.

**Comment by richardstartin** [comment]
This should probably allow for CASing array elements too.

**Pull request closed without merging**